vtm/src/org/oscim/tiling/TileSource.java

```diff
@@ -146,7 +151,25 @@ public int getOverZoom() {
     }
 
     public void setFadeSteps(FadeStep[] fadeSteps) {
-        mFadeSteps = fadeSteps;
+        if (fadeSteps == null) {
+            mFadeSteps = null;
+            return;
+        }
+        // Arrays.asList not support add, as it's read only.
+        List<FadeStep> fadeStepsList = new ArrayList<>(Arrays.asList(fadeSteps));
+        Collections.sort(fadeStepsList);
+        for (int i = 1; i < fadeSteps.length; i++) {
+            FadeStep step = fadeStepsList.get(i - 1);
+            FadeStep next = fadeStepsList.get(i);
+            if (step.zoomEnd > next.zoomStart) {
+                throw new IllegalArgumentException("FadeSteps interfere with each other" + getClass().getName());
+            } else if (step.zoomEnd < next.zoomStart) {
+                // Add new FadeStep transition if there's a gap
+                fadeStepsList.add(new FadeStep(step.scaleEnd, next.scaleStart, step.alphaEnd, next.alphaStart));
+            }
+        }
+        Collections.sort(fadeStepsList);
+        mFadeSteps = fadeStepsList.toArray(new FadeStep[0]);
```

Comment:
What is the purpose for that in more detail?
Better let user handle own configuration without the library trying to automate / interpolate.
e.g. some may want fade steps to be distant from each other with gaps for some purpose.

Reply:
Reason is that having one FadeStep only, user expects that it takes the max / min values. But how to handle cases in between two FadeSteps then?

Reply:
Users create the fade steps, so they know their inputs, without library assumptions.
Are there any cases which need to be resolved?
There are some fade steps examples in [DefaultSources](https://github.com/mapsforge/vtm/blob/master/vtm/src/org/oscim/tiling/source/bitmap/DefaultSources.java) which seem working well.

Reply:
> But how to handle cases in between two FadeSteps then?

Doesn't seem anything to handle here, from library's perspective.
If users want having distant fade steps, library should respect that.
And if want adjacent fade steps, could create them from the start.

Reply:
My concern was, if having only on fade e.g. from 17 to 18 to fade a layer to alpha 0. Then I expected that beyond this zoom levels it remains on alpha 0. But that approach doesn't seem to please you :wink: so can leave it.

Reply:
> Then I expected that beyond this zoom levels it remains on alpha 0.

That's an assumption, other users may expect (and should have the option) to work differently.
Now there is plenty flexibility, could just add another fade step for the needed alpha 0 & zooms.

> But that approach doesn't seem to please you so can leave it.

It's always about library and its stable behavior for all users. :slightly_smiling_face: 

_Note: to not lose time in both implementation and reviewing, such changes can be discussed first (e.g. in forum) to see where / how / if there is something to improve and change._

Reply:
_Sometimes I make these changes while doing other stuff and need it, so for me it's more comfortable to show these changes, than explaining them and end up in a bunch of misapprehensions. And I only make them, if I think they could help others. But I understand your concern._